power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,50 +766,113 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+            }
+            if (result.is_columnar) {
+                process_buffer_span.template operator()<CT>(
+                    update_data.get_columnar_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
+            } else {
+                process_buffer_span.template operator()<CT>(
+                    update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
             }
-            return process_buffer_span.template operator()<CT>(
-                update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>());
+            return result;
         };
 
-        // check all components
-        auto const update_independent = run_functor_with_all_types_return_array(is_component_update_independent);
-        return std::ranges::all_of(update_independent, [](bool const is_independent) { return is_independent; });
+        // check and return indenpendence of all components
+        return run_functor_with_all_types_return_vector(check_each_component);
+    }
+
+    static bool is_update_independent(ConstDataset const& update_data) {
+        // If the batch size is (0 or) 1, then the update data for this component is 'independent'
+        if (update_data.batch_size() <= 1) {
+            return true;
+        }
+        auto const all_comp_update_independence = check_components_independence(update_data);
+        return std::ranges::all_of(all_comp_update_independence,
+                                   [](auto const& comp) { return !comp.has_id || comp.ids_match; });
+    }
+
+    void validate_update_data_independence(UpdateCompProperties const& comp) const {
+        if (!comp.has_id && comp.ids_all_na) {
+            return; // empty dataset is still supported
+        }
+        if (comp.elements_ps_in_base < comp.elements_ps_in_update) {
+            throw DatasetError("Update data has more elements per scenario than base data for component " + comp.name +
+                               "!");
+        }
+        if (comp.ids_part_na) {
+            throw DatasetError("Part of the IDs are not valid for component " + comp.name + " in update data!");
+        }
+        if (!comp.uniform) {
+            if (comp.is_columnar && !comp.has_id) {
+                throw DatasetError("Columnar input data without IDs for component " + comp.name + " is not uniform!");
+            }
+            if (!comp.is_columnar && comp.ids_all_na) {
+                throw DatasetError("Row based input data with all NA IDs for component " + comp.name +
+                                   " is not uniform!");
+            }
+        }
+        if (comp.elements_ps_in_base != comp.elements_ps_in_update) {
+            if (comp.is_columnar && !comp.has_id) {
+                throw DatasetError("Columnar input data for component " + comp.name +
+                                   " has different number of elements per scenario in update and base data!");
```

Comment:
Now I am a bit confused about base data. Isn't base data just input data? Perhaps its me, but maybe a bit of rephrasing here would be good.

Reply:
base data = data instance living inside main model; input data = data used to feed to main model to 'create' base data. They are the same data just in different lives.

Reply:
Since the user isn't supposed to know what base data is, then maybe some re-phrasing here could be a good idea.